backend/src/app/core/authorization/authz_user.py

```diff
@@ -34,31 +36,61 @@ def __init__(
         self.user = user
         self.db = db
 
-    def assert_in_same_project_as(self, crud: Crud, object_id: int):
+    def assert_in_same_project_as(self, crud: Crud, object_id: int | str):
         try:
-            orm_object = crud.value.read(self.db, object_id)
+            # Some read functions take an int, others take a str
+            # (e.g. PreprocessingJobs).
+            # There's no good way to bind the value of `Crud` to the
+            # type of `object_id`, so we switch off the type checker
+            # for this line :(
+            orm_object = crud.value.read(self.db, object_id)  # type: ignore
         except NoSuchElementError:
-            self.assert_bool(False, "Object does not exist")
-            return
+            self.deny_access(f"{crud.name} {object_id} does not exist")
 
         project_id = get_parent_project_id(orm_object)
         if project_id is None:
-            self.assert_bool(False, "Object has no parent project")
-            return
+            self.deny_access("Object has no parent project")
 
         self.assert_in_project(project_id)
 
-    def assert_object_has_same_user_id(self, crud: Crud, object_id: int):
+    def assert_in_same_project_as_many(
+        self, crud: Crud, object_ids: List[int] | List[str]
+    ):
+        # Some read functions take an int, others take a str
+        # (e.g. PreprocessingJobs).
+        # There's no good way to bind the value of `Crud` to the
+        # type of `object_id`, so we switch off the type checker
+        # for this line :(
+        objs = crud.value.read_by_ids(self.db, object_ids)  # type: ignore
```

Comment:
to make this more efficient, we could create a `exists_by_ids` method that relies on a single SQL query

Reply:
I've replaced the for loop at the bottom of this method with a single query, but I think we still need to read all these objects into memory to find their project ids, since it would be quite the effort to convert `get_parent_project_id` into an SQL query :thinking: